ludwig/data/prompt.py

```diff
@@ -169,44 +174,64 @@ def format_input_with_prompt(
             template = DEFAULT_ZERO_SHOT_PROMPT_TEMPLATE
 
     # ensure that the prompt template has all required fields
+    template_fields, field_to_dtype = _get_template_fields(template)
     try:
-        _validate_prompt_template(template, is_few_shot=is_few_shot)
+        _validate_prompt_template(template_fields, task_str, is_few_shot, dataset_df.columns, input_col_name)
     except ValueError as e:
         raise ValueError(f"template invalid for {'few-shot' if is_few_shot else 'zero-shot'} prompt: {e}")
 
-    def generate_prompt(series: pd.Series):
-        df = series.to_frame(name=input_col_name)
+    def generate_prompt(df: pd.DataFrame):
+        if CONTEXT in template_fields:
+            df[CONTEXT] = retrieval_model.search(df, backend, k=k, return_data=True)
+        if SAMPLE in template_fields:
+            df[SAMPLE] = df[input_col_name].map(lambda entry: json.dumps(entry, indent=2))
+        if TASK in template_fields:
+            df[TASK] = task_str
 
-        if is_few_shot:
-            df["context"] = retrieval_model.search(df, backend, k=k, return_data=True)
+        def generate_prompt_for_row(row):
+            kwargs = {col: field_to_dtype[col](row[col]) for col in template_fields}
```

Comment:
why do we need to cast the values?

Reply:
If you have a number like `0.1234` in the DF and you want to format it like `{number:.2f}` it will fail if the number is represented as a string in the DF. So we need to cast it to a float. I can add a comment to this effect.